spirv-tools/src/cmd.rs

```diff
@@ -0,0 +1,192 @@
+use crate::error::Message;
+use std::process::{Command, Stdio};
+
+pub enum CmdError {
+    /// The binary failed to spawn, probably because it's not installed
+    /// or not in PATH
+    BinaryNotFound(std::io::Error),
+    /// An I/O error occurred accessing the process' pipes
+    Io(std::io::Error),
+    /// The binary ran, but returned a non-zero exit code and (hopefully)
+    /// diagnostics
+    ToolErrors {
+        exit_code: i32,
+        /// Messages that were parsed from the output
+        messages: Vec<Message>,
+    },
+}
+
+impl From<CmdError> for crate::error::Error {
+    fn from(ce: CmdError) -> Self {
+        use crate::SpirvResult;
+
+        match ce {
+            CmdError::BinaryNotFound(e) => Self {
+                inner: SpirvResult::Unsupported,
+                diagnostic: Some(format!("failed spawn executable: {}", e).into()),
+            },
+            CmdError::Io(e) => Self {
+                inner: SpirvResult::EndOfStream,
+                diagnostic: Some(
+                    format!("i/o error occurred communicating with executable: {}", e).into(),
+                ),
+            },
+            CmdError::ToolErrors {
+                exit_code: _,
+                messages,
+            } => {
+                // The C API just puts the last message as the diagnostic, so just do the
+                // same for now
+                let diagnostic = messages
+                    .into_iter()
+                    .last()
+                    .map(crate::error::Diagnostic::from);
+
+                Self {
+                    inner: SpirvResult::InternalError, // this isn't really correct
+                    diagnostic,
+                }
+            }
+        }
+    }
+}
+
+pub struct CmdOutput {
+    /// The output the command is actually supposed to give back
+    pub binary: Vec<u8>,
+    /// Warning or Info level diagnostics that were gathered during execution
+    pub messages: Vec<Message>,
+}
+
+#[derive(PartialEq, Copy, Clone)]
+pub enum Output {
+    /// Doesn't try to read stdout for tool output (other than diagnostics)
+    Ignore,
+    /// Attempts to retrieve the tool's output from stdout
+    Retrieve,
+}
+
+pub fn exec(
+    mut cmd: Command,
+    input: Option<&[u8]>,
+    retrieve_output: Output,
+) -> Result<CmdOutput, CmdError> {
+    if input.is_some() {
+        cmd.stdin(Stdio::piped());
+    }
+
+    cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
+
+    let mut child = cmd.spawn().map_err(CmdError::BinaryNotFound)?;
+
+    if let Some(input) = input {
+        use std::io::Write;
+
+        child
+            .stdin
+            .take()
+            .unwrap()
+            .write_all(input)
```

Comment:
oh wow, I definitely wouldn't have thought of piping the input, this is super cool!

Reply:
Heh yes, a bit nicer than writing a temporary file. :slightly_smiling_face: